8253266 : JList and JTable constructors should clear OPAQUE_SET before calling updateUI

trebari opened the pull request:
Hi All,
Please review the following fix for jdk17.

Issue : LookAndFeel.installProperty(list, "opaque", false) is not able to set the opaque property for JList and JTable.
LookAndFeel.installProperty calls the setUIProperty, and setUIProperty checks for OPAQUE_SET to change the opaque property as requested by the client.
OPAQUE_SET is always set to true when there is call to setOpaque(boolean).So when the constructor calls setOpaque(true) OPAQUE_SET is set to true and wont allow the setUIProperty to change the opaque property.
installProperty should work as the opaque property is not set by the client.

Fix. Fix is to remove the call to the setOpaque() from the constructor of JList and JTable. This will allow the client to change the opaque property calling LookAndFeel.installProperty() when the property is already not set.

Test : Added a  test  to check the same. Also tested internal tests which all are passing.
Link is in JBS.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8253266](https://bugs.openjdk.java.net/browse/JDK-8253266): JList and JTable constructors should clear OPAQUE_SET before calling updateUI


### Reviewers
 * [Prasanta Sadhukhan](https://openjdk.java.net/census#psadhukhan) (@prsadhuk - **Reviewer**)
 * [Sergey Bylokhov](https://openjdk.java.net/census#serb) (@mrserb - **Reviewer**) ⚠️ Review applies to 67fbef160e7715e8a4ba30c068a37cc97a637fef


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/3167/head:pull/3167` \
`$ git checkout pull/3167`

Update a local copy of the PR: \
`$ git checkout pull/3167` \
`$ git pull https://git.openjdk.java.net/jdk pull/3167/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 3167`

View PR using the GUI difftool: \
`$ git pr show -t 3167`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/3167.diff">https://git.openjdk.java.net/jdk/pull/3167.diff</a>

</details>

prsadhuk reviewed (commented):
I am not sure if this is the correct fix, as then we would have to remove setOpaque call for JTree, JToolTip, JRootPane etc where we might face same installProperty issue. 
Also, if we have to change this, we then need to change "autoScrolls" property too which is set in this 2 component and setUIProperty() can change those too.
Probably, we need to check if the value to be set is different in setUIProperty and then set it 
or 
override setUIProperty per component basis (as it is done for JTable) and then not check for OPAQUE_SET flag for opaque property accordingly.

trebari commented:
> I am not sure if this is the correct fix, as then we would have to remove setOpaque call for JTree, JToolTip, JRootPane etc where we might face same installProperty issue.

Yes we face the same issue in JTree, JToolTip, and JViewport , 
but many of the components do not  have this problem like JButton JRootPane , JCheckbox, JInternalFrame, JMenuItem, JCheckBoxMenuItem, JPopupMenu, JMenu ..etc
We can fix the issue with JTree, JTooltip and JViewport same way.

> Also, if we have to change this, we then need to change "autoScrolls" property too which is set in this 2 component and setUIProperty() can change those too.
> Probably, we need to check if the value to be set is different in setUIProperty and then set it 
> or
> override setUIProperty per component basis (as it is done for JTable) and then not check for OPAQUE_SET flag for opaque property accordingly.

OPAQUE_SET is private variable and can not be accessed in JTable or JTree. 
From the spec of LookAndFeel.installProperty and setUIProperty it
  * Sets the property with the specified name to the specified value if
     * the property has not already been set by the client program.*
 is clear that once the property is set by the client setUIProperty can not change it. And OPAUQE_SET  is  used  for  that
.

Removing the setOpaque() from the constructor of these component will not make any initial difference as it is also set to true in BasicListUI, BasicTableUI, BasicTreeUI, BasicTooltipUI and BasicViewportUI by calling LookAndFeel.installProperty.
The difference which it will make is now by calling LookAndFeel.installProperty(c,opaque,false) will clear the opaque property.

mrserb commented on test/jdk/javax/swing/JList/TestOpaqueListTable.java:
```diff
@@ -0,0 +1,81 @@
+public class TestOpaqueListTable {
+
+    public static void main(String[] args) throws Exception {
+        UIManager.LookAndFeelInfo[] installedLookAndFeels;
```
I did not look close to the fix, but you need to check that the default value of the "opaque" property will not be changed. So all default L&F will set it to true, as before the fix.

trebari replied:
Yeah, the default value for "opaque" property will not be changed.
It will be true for All L&F after the fix as well , I will add  that check to the test.

trebari commented:
@prsadhuk  @mrserb is there any more comments on this.

mrserb reviewed (approved):
Please wait for the @prsadhuk approval as well.

trebari commented:
I have fixed the issue with JTree, JTooltip and JViewport same way.
While fixing for JTooltip discovered that default value for JToolTip opaque property for Nimbus LookAndFeel is set to false in the skin.laf file. This value was never taken into account because the setOpaque(true) from the JTooltip was overriding this.
As we are now removing it from the constructor of JTooptip, NimbusLAF is setting to the false.
To keep the behaviour same as before i have set it to true for NimbusLAF.

Also modified the test to check for JTooltip , JTree and JViewport.

trebari commented:
Internal testing looks good, link is in JBS

prsadhuk commented on src/java.desktop/share/classes/javax/swing/plaf/nimbus/skin.laf:
```diff
@@ -27248,7 +27248,7 @@
          <regions/>
       </uiComponent>
-      <uiComponent opaque="false" type="javax.swing.JToolTip" name="ToolTip" ui="ToolTipUI" subregion="false">
+      <uiComponent opaque="true" type="javax.swing.JToolTip" name="ToolTip" ui="ToolTipUI" subregion="false">
```
In my opinion, instead of changing skin.laf we probably should update SynthToolTipUI.installDefaults() and add 
`LookAndFeel.installProperty(c, "opaque", Boolean.TRUE);`

similar to what we do for BasicToolTipUI.installDefaults. Will it not work?

trebari replied:
No, it doesn't work, i have checked.
The change needs to be skin.laf file only.

prsadhuk replied:
I am not sure where you placed the code. I think it's need to be added after updateStyle() as updateStyle() uninstall the default and install for new style.
```
protected void installDefaults(JComponent c) {
        updateStyle(c);
    }
```

trebari replied:
yeah setting the property after updateStyle(c) works fine.
updating the change in the SynthToolTipUI.installDefaults()

mrserb replied:
What is the difference from setting it via "skin.laf"? The "skin.laf" is converted to the NimbusDefaults during the build and these defaults are used by the NimbusLookAndFeel.

prsadhuk replied:
I personally feel setting in code will be easier to maintain going forward than to go through property file settings.

mrserb replied:
Then why we need this file at all? Now we will set the false value in the config, and true at runtime, what is the profit?

mrserb replied:
I think we should set it in the config file since this is the place where all other properties are stored.

prsadhuk replied:
Before this fix, I see it was set in property file and then overridden in JToolTip. Current fix does the same, set in property file and then overridden in SynthToolTipUI. But I don't have objection if it is only controlled by property file.

prsadhuk commented on test/jdk/javax/swing/JList/TestOpaqueListTable.java:
```diff
@@ -0,0 +1,147 @@
+ * questions.
+ */
+
+import javax.swing.*;
```
Can you please update this wildcard imports?

trebari replied:
Done

prsadhuk commented on test/jdk/javax/swing/JList/TestOpaqueListTable.java:
```diff
@@ -0,0 +1,147 @@
+ *  @bug 8253266
+ *  @summary setUIProperty should work when opaque property is not set by
+ *  client
+ *  @key headful
```
Does it need to be headful?

trebari replied:
It is better to run this test on headful systems, which has the required components installed.
This test checks for all look and feel. I remember making one of the non ui test as not headful and it thrown LookAndFeelNot installed exception.
So one of the suggestion was to make it headful, so doing the same here.

prsadhuk commented on src/java.desktop/share/classes/javax/swing/plaf/nimbus/skin.laf:
```diff
@@ -21578,7 +21578,7 @@
                            <shapes>
                               <!--<rectangle x1="3.9545454545454546" x2="6.0" y1="26.49999999999998" y2="10.772727272727275" rounding="2.000000000000001">-->
-                                 
+
```
Is this newline needed?

trebari replied:
No , it automatically got added.Need to remove this.

trebari replied:
Done

trebari commented:
/integrate
